pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/client/Auth0ManagementClient.java

```diff
@@ -538,15 +540,25 @@ private <B, E> ApiResult<B, E> withRetries(String retryMessage, Supplier<ApiResu
                 break;
             }
             if (res.getStatusCode() == 429) {
-                log.error(retryMessage, res.getError());
+                log.error(retryMessage, res.getError() + " " + res.getError().getClass().getName());
                 long wait = backoffMillis * numTries + new Random().nextInt(MAX_JITTER_MILLIS);
+                // if we have more information from auth0 about how long to wait, use it.
+                if (res.getError() instanceof RateLimitException) {
+                    RateLimitException rateLimit = (RateLimitException) res.getError();
+                    long unixTimeAtWhichToRetry = rateLimit.getReset();
+                    long suggestedWaitTime = unixTimeAtWhichToRetry - Instant.now().getEpochSecond();
+                    if (suggestedWaitTime > 0) {
+                        wait = suggestedWaitTime * 1000;
+                    }
+                    log.warn("Hit auth0 rate limit.  Pausing for " + wait + "s based on auth0 headers.");
+                } else {
+                    log.warn("Hit auth0 rate limit.  Pausing for " + wait + "ms");
+                }
                 try {
                     TimeUnit.MILLISECONDS.sleep(wait);
                 } catch (InterruptedException e) {
                     log.warn("Interrupted while waiting after rate limit", e);
                 }
-            } else {
-                break;
```

Comment:
Fixing a bug in snippet:
```
else { break; }
```
is wrong. It causes function to exit early before all retries has completed.